pir/cpp/database.cpp

```diff
@@ -44,90 +48,132 @@ StatusOr<std::shared_ptr<PIRDatabase>> PIRDatabase::Create(
+   */
+  DatabaseMultiplier(const vector<Plaintext>& database,
+                     const vector<Ciphertext>& selection_vector,
+                     Evaluator& evaluator,
+                     const seal::RelinKeys* const relin_keys,
+                     seal::Decryptor* const decryptor)
+      : database_(database),
+        selection_vector_(selection_vector),
+        evaluator_(evaluator),
+        relin_keys_(relin_keys),
+        decryptor_(decryptor) {}
+
+  /**
+   * Do the multiplication using the given dimension sizes.
+   */
+  Ciphertext multiply(vector<uint32_t> dimensions) {
+    database_it_ = database_.begin();
+    return multiply(dimensions, selection_vector_.begin(), 0);
+  }
 
-      if (relin_keys != nullptr) {
-        evaluator.relinearize_inplace(temp_ct, *relin_keys);
-        if (decryptor != nullptr) {
-          std::cout << depth_string << "i = " << i << " noise budget relin "
-                    << decryptor->invariant_noise_budget(temp_ct) << std::endl;
+ private:
+  /**
+   * Recursive function to do the dot product of each dimension with the db.
+   * Calls itself to move down dimensions until you get to the bottom dimension.
+   * Bottom dimension just does a dot product with the DB, and returns the
+   * result. Upper levels then take those results, and dot product again with
+   * the selection vector, until you get back to the top. NB: Database iterator
+   * is kept at the class level so that we move through the database one element
+   * at a time.
+   *
+   * @param[in] dimensions List of remaining demainsion sizes.
+   * @param[in] selection_vector_it Iterator into the start of the selection
+   *  vector for the current depth.
+   * @param[in] depth Current depth.
+   */
+  Ciphertext multiply(vector<uint32_t> dimensions,
+                      vector<Ciphertext>::const_iterator selection_vector_it,
+                      size_t depth) {
+    const size_t this_dimension = dimensions[0];
+    auto remaining_dimensions =
+        vector<uint32_t>(dimensions.begin() + 1, dimensions.end());
+
+    string depth_string(depth, ' ');
+
+    Ciphertext result;
+    bool first_pass = true;
+    for (size_t i = 0; i < this_dimension; ++i) {
+      // make sure we don't go past end of DB
+      if (database_it_ == database_.end()) break;
+      Ciphertext temp_ct;
+      if (remaining_dimensions.empty()) {
+        // base case: have to multiply against DB
+        evaluator_.multiply_plain(*(selection_vector_it + i), *(database_it_++),
+                                  temp_ct);
+        print_noise(depth, i, "base", temp_ct);
+
+      } else {
+        temp_ct = multiply(remaining_dimensions,
+                           selection_vector_it + this_dimension, depth + 1);
+        print_noise(depth, i, "recurse", temp_ct);
+
+        evaluator_.multiply_inplace(temp_ct, *(selection_vector_it + i));
+        print_noise(depth, i, "mult", temp_ct);
+
+        if (relin_keys_ != nullptr) {
```

Comment:
Is there a scenario where we should proceed here without the relin_keys? 
Won't the ciphertext get huge without them?

Reply:
Definitely need them at depth >= 3, at 2 they're a good idea but I don't think they're essential. I could add a check that at depth >= 2 they have to be present

Reply:
Thinking more about this, again I think you're right, but I don't think that should be in this PR. I've made #28 for this decision.